### PR TITLE
Fix annotations preservation

### LINKED ISSUE #1134: If annotating case classes with BigQueryType other static macro annotations aren't preserved
In our investigations for adding a new macro annotation we encountered a Scio bug.  When a user adds a static macro annotation such as `@BigQueryType.toTable` we noticed that any other annotations are removed if they exist below this one.

An example is

```scala
@BigQueryType.toTable
@Test.myMacroAnnotation
case class TestClass
```

This will result in only the first macro being triggered with the second one being ignored.  This is because macros are processed in order.  Since the actual ordering matters and each successive static macro annotation can change the Abstract Syntax Tree if we leave any other static macro annotations from what is produced by a given macro the other ones are never triggered at all.

In our situation this is actually not that big of a deal.  We can require people to put our macro above `@BigQueryType.toTable` and everything still triggers as expected but this seems like a bug in Scio and could lead to user confusion at some point.

Full Example:

```scala
import scala.annotation.StaticAnnotation
  import scala.language.experimental.macros
  import scala.reflect.macros._

  def validateImpl(c: blackbox.Context)(annottees: c.Expr[Any]*): c.Expr[Any] = {
    println("Testing")
    annottees.head
  }
  class bigQueryTypeConstruct extends StaticAnnotation {
    def macroTransform(annottees: Any*): Any = macro validateImpl
  }

  @BigQueryType.toTable
  @bigQueryTypeConstruct
  case class TestNoPrint()

  @bigQueryTypeConstruct
  @BigQueryType.toTable
  case class TestPrints()
```

Note that testing only prints on the second example.

Set `sys.props("bigquery.types.debug") = "true"` in order to see the full case classes generated while macros expand.

### PR DESCRIPTION
**Pull request by oskoi**
Fixes #1134

### CODE REVIEW

**Comment by nevillelyh** [comment]
@jbigred1 PTAL

File: scio-bigquery/src/main/scala/com/spotify/scio/bigquery/types/TypeProvider.scala [hunk 1]
```diff
@@ -87,6 +86,7 @@ private[types] object TypeProvider {
         if (parents.map(_.toString()).toSet != Set("scala.Product", "scala.Serializable")) {
           c.abort(c.enclosingPosition, s"Invalid annotation, don't extend the case class $clazzDef")
         }
+
```
Comment (jbigred1) [inline scio-bigquery/src/main/scala/com/spotify/scio/bigquery/types/TypeProvider.scala]: unnecessary new line addition/removal 😄

File: scio-bigquery/src/test/scala/com/spotify/scio/bigquery/types/TypeProviderTest.scala [hunk 2]
```diff
@@ -450,4 +453,30 @@ class TypeProviderTest extends FlatSpec with Matchers {
...
+  }
+
+  @UserDefinedAnnotation1
+  @BigQueryType.fromSchema(
+    """{"fields": [ {"mode": "REQUIRED", "name": "f1", "type": "INTEGER"} ]}""")
+  @UserDefinedAnnotation2
+  class ClassWithCustomAnnotations
+
+  "BigQueryType.fromSchema" should "support user defined annotations" in {
+    val annotations = typeOf[ClassWithCustomAnnotations].typeSymbol.annotations.map(_.tree.tpe)
```
Comment (jbigred1) [inline scio-bigquery/src/test/scala/com/spotify/scio/bigquery/types/TypeProviderTest.scala]: you can factor this logic into a separate function that takes a type to test

something like `def containsAllTypes[T]` and then test it using `T` instead

File: scio-bigquery/src/test/scala/com/spotify/scio/bigquery/types/TypeProviderTest.scala [hunk 3]
```diff
@@ -450,4 +453,30 @@ class TypeProviderTest extends FlatSpec with Matchers {
...
+    annotations should contain allOf(
+      typeOf[UserDefinedAnnotation1], typeOf[UserDefinedAnnotation2])
+  }
+
+  @UserDefinedAnnotation1
+  @BigQueryType.fromSchema(
+    """{"fields": [ {"mode": "REQUIRED", "name": "f1", "type": "INTEGER"} ]}""")
+  @UserDefinedAnnotation2
+  class ClassWithCustomAnnotations
+
```
Comment (jbigred1) [inline scio-bigquery/src/test/scala/com/spotify/scio/bigquery/types/TypeProviderTest.scala]: do you think it's worthwhile testing with other macro annotations that change the AST or is this sufficient

File: scio-bigquery/src/main/scala/com/spotify/scio/bigquery/types/TypeProvider.scala [hunk 4]
```diff
@@ -229,9 +230,11 @@ private[types] object TypeProvider {
 
   /** Generate a case class. */
   private def caseClass(c: blackbox.Context)
-                       (name: c.TypeName, fields: Seq[c.Tree], body: Seq[c.Tree]): c.Tree = {
+                       (mods: c.Modifiers, name: c.TypeName, fields: Seq[c.Tree], body: Seq[c.Tree]): c.Tree = {
     import c.universe._
-    q"@_root_.com.spotify.scio.bigquery.types.BigQueryTag case class $name(..$fields) extends ${p(c, SType)}.HasAnnotation { ..$body }"
+    val tagAnnot = q"new _root_.com.spotify.scio.bigquery.types.BigQueryTag"
+    val taggedMods = Modifiers(Flag.CASE, typeNames.EMPTY, tagAnnot :: mods.annotations)
+    q"$taggedMods class $name(..$fields) extends ${p(c, SType)}.HasAnnotation { ..$body }"
```
Comment (jbigred1) [inline scio-bigquery/src/main/scala/com/spotify/scio/bigquery/types/TypeProvider.scala]: do we need the `case` modifier or is that what the first argument to this function does

**Comment by jbigred1** [comment]
any idea why the tests are having issues on `2.11.x` versions now

**Comment by jbigred1** [comment]
aside from those small comments it looks good

**Comment by oskoi** [comment]
@jbigred1 I had to reduce the size of Avro TypeProviderTest file a little  because of the limitations (800 lines per file).
All tests passed. Make a review, please)

File: scio-avro/src/test/scala/com/spotify/scio/avro/types/TypeProviderTest.scala [hunk 5]
```diff
@@ -21,45 +21,39 @@ import com.google.protobuf.ByteString
...
-      {"type":"record","name": "Record","fields":[{"name":"f1","type":"int"}]}
-    """)
+    """{"type":"record","name": "Record","fields":[{"name":"f1","type":"int"}]}""")
   class MultiLineStringLiteral
 
   @AvroType.fromSchema(
-    """
-      |{"type":"record","name": "Record","fields":[{"name":"f1","type":"int"}]}
-    """.stripMargin)
+    """{"type":"record","name": "Record","fields":[{"name":"f1","type":"int"}]}""")
```
Comment (jbigred1) [inline scio-avro/src/test/scala/com/spotify/scio/avro/types/TypeProviderTest.scala]: what's the reason to remove the `stripMargin`

File: scio-avro/src/test/scala/com/spotify/scio/avro/types/TypeProviderTest.scala [hunk 6]
```diff
@@ -763,13 +754,45 @@ class TypeProviderTest extends FlatSpec with Matchers {
...
+  @UserAnnot2
+  class SchemaWithUserAnnotations
+
+  "AvroType.fromSchema" should "preserve user defined annotations" in {
+    containsAllUserAnnotTypes[SchemaWithUserAnnotations]
+  }
+
+  @UserAnnot1
+  @AvroType.toSchema
+  @UserAnnot2
```
Comment (jbigred1) [inline scio-avro/src/test/scala/com/spotify/scio/avro/types/TypeProviderTest.scala]: can you test different orderings as well

File: scio-avro/src/test/scala/com/spotify/scio/avro/types/TypeProviderTest.scala (hunk 5 above)
Comment (oskoi) [reply scio-avro/src/test/scala/com/spotify/scio/avro/types/TypeProviderTest.scala]: The length of the file went beyond the permissible limits (800 lines per file). I had to go for such tricks. If this is critical, I can return. What do you think?

File: scio-avro/src/test/scala/com/spotify/scio/avro/types/TypeProviderTest.scala (hunk 6 above)
Comment (oskoi) [reply scio-avro/src/test/scala/com/spotify/scio/avro/types/TypeProviderTest.scala]: The length of the file exceeds the permissible limit. Where it is better to add tests with annotations?

File: scio-avro/src/test/scala/com/spotify/scio/avro/types/TypeProviderTest.scala (hunk 5 above)
Comment (jbigred1) [reply scio-avro/src/test/scala/com/spotify/scio/avro/types/TypeProviderTest.scala]: @nevillelyh what do you think

that seems like a silly reason to remove stylistic elements and things that we're testing

File: scio-avro/src/test/scala/com/spotify/scio/avro/types/TypeProviderTest.scala (hunk 6 above)
Comment (jbigred1) [reply scio-avro/src/test/scala/com/spotify/scio/avro/types/TypeProviderTest.scala]: could be time to add another file to specifically test annotations

is it possible to increase the file size for tests @nevillelyh what do you think

Comment (nevillelyh) [reply scio-avro/src/test/scala/com/spotify/scio/avro/types/TypeProviderTest.scala]: We can just add a pair of `scalastyle:{off,on}` to suppress the limit like this: https://github.com/spotify/scio/blob/master/scio-core/src/main/scala/com/spotify/scio/ScioContext.scala#L18

File: scio-avro/src/test/scala/com/spotify/scio/avro/types/TypeProviderTest.scala (hunk 5 above)
Comment (nevillelyh) [reply scio-avro/src/test/scala/com/spotify/scio/avro/types/TypeProviderTest.scala]: You can also surround this with a pair of `// scalastyle:{off,on} line.size.limit` to suppress the limit.

**Comment by nevillelyh** [comment]
Same, LGTM except the minor comments. Prefer suppressing those style warnings than working around them.

**Comment by oskoi** [comment]
@nevillelyh Oh, thanks for the advice!
@jbigred1 Done. Reverted old changes and added test cases with different annotations ordering

**Comment by jbigred1** [comment]
this looks good

@nevillelyh can you merge it when you have a second

**Comment by jbigred1** [comment]
also thanks for fixing the Avro part as well

**Pull request merged**